### PR TITLE
feat(auth): ensure customer currency is set when needed

### LINKED ISSUE #12301: Some existing PayPal customers' `currency` is `null` in Firestore
[https://mozilla-hub.atlassian.net/browse/FXA-4779](https://mozilla-hub.atlassian.net/browse/FXA-4779|smart-link) should fix this issue for new PayPal customers, but we should update existing PayPal customers' `currency` in Firestore, as they would have the same issue upgrading if they ever tried.

I plan to ask our SRE folks to search our `fxa-prod` Firestore DB to see how many customers are affected, as if it’s a small number, we could manually trigger a `customer.updated` webhook through the Stripe dashboard. Otherwise we will need to write and run a script to update these customer records.

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/FXA-4784)

### PR DESCRIPTION
**Pull request by bbangert**
Because:

* Due to a race condition on paypal customers, it was possible that
  we did not get the latest customer record with currency when
  creating the billing agreement and subscription.

This commit:

* Updates fetchCustomer to always refresh the customer if their
  record is missing currency but they have subscriptions.
* Updates the stripe webhook to refresh as needed on subscription
  events if no currency is detected.

Closes #12301

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 1]
```diff
@@ -3154,18 +3165,18 @@ export class StripeHelper {
...
-    // an unexpanded subscription object, we update all customers in Firestore.
-    if (event.type === 'customer.subscription.created') {
+    // Update the customer if our copy of the customer is missing the currency.
+    // This could occur in some edge cases where the subscription is created
+    // before a user paying with paypal has the paypal agreement ID set on the
+    // user.
+    const customer = await this.expandResource(
+      subscription.customer,
+      CUSTOMER_RESOURCE
+    );
```
Comment (biancadanforth) [inline packages/fxa-auth-server/lib/payments/stripe.ts:3175]: The reason why I didn't previously fetch the customer from Stripe is that I know `fetchAndInsertCustomer` does that fetch already.

With the previous solution, we'd update our cached customer/subscriptions for all customers (Stripe and PayPal) on every `customer.subscription.updated` webhook but only make one Stripe call.

With this solution, we only update the cache in the applicable case, but we would also make a second Stripe call in the applicable case.

I think your solution is cleaner and more clear why we're doing this; just thinking about the trade-offs.

File: packages/fxa-auth-server/test/local/payments/stripe.js [hunk 2]
```diff
@@ -3353,6 +3353,25 @@ describe('StripeHelper', () => {
...
+      customer.currency = null;
+      const customerSecond = deepCopy(customer1);
+      const expandStub = sandbox.stub(stripeHelper, 'expandResource');
+      stripeHelper.stripeFirestore = {
+        fetchAndInsertCustomer: sandbox.stub().resolves({}),
+      };
+      expandStub.onFirstCall().resolves(customer);
+      expandStub.onSecondCall().resolves(customerSecond);
+      const result = await stripeHelper.fetchCustomer(existingCustomer.uid);
+      assert.deepEqual(result, customer1);
```
Comment (biancadanforth) [inline packages/fxa-auth-server/test/local/payments/stripe.js:3367]: For clarity, should we assert `result` deep equals `customerSecond`? And maybe we can rename `customerSecond` to `customerWithCurrency`?

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 3]
```diff
@@ -3154,18 +3165,18 @@ export class StripeHelper {
     const subscription = await this.stripe.subscriptions.retrieve(
       (event.data.object as Stripe.Subscription).id
     );
-    // For PayPal customers, Stripe fails to send a `customer.updated` webhook
-    // when it automatically updates the `customer.currency` on subscription
-    // creation. Update the cached customer here to ensure `customer.currency`
-    // is not `null`.
-    // Since we can't easily differentiate a PayPal customer from others with
-    // an unexpanded subscription object, we update all customers in Firestore.
-    if (event.type === 'customer.subscription.created') {
```
Comment (biancadanforth) [inline packages/fxa-auth-server/lib/payments/stripe.ts:3163]: Why do we also want to do this for `customer.subscription.updated`? Is this to catch and fix the issue eagerly e.g. when `latest_invoice` changes on a renewal?

**Review by biancadanforth — APPROVED** [review]
LGTM except I'm not sure we need to call `expandResource` after `fetchAndInsertCustomer` in `fetchCustomer`, so perhaps that can be removed?

I also made one or two minor questions and suggestions.

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 4]
```diff
@@ -1534,6 +1534,17 @@ export class StripeHelper {
...
+    // If the customer has subscriptions and no currency, we must have a stale
+    // customer record. Let's update it.
+    if (customer.subscriptions?.data.length && !customer.currency) {
+      await this.stripeFirestore.fetchAndInsertCustomer(customer.id);
+      // Retrieve the customer again.
+      customer = await this.expandResource<Stripe.Customer>(
+        stripeCustomerId,
+        CUSTOMER_RESOURCE
+      );
+    }
```
[the remark this answers is not in the document]
Comment (bbangert) [reply packages/fxa-auth-server/lib/payments/stripe.ts:1546]: The customer returned by this method does not have the expanded subscriptions on it, while the other method does add it.

File: packages/fxa-auth-server/lib/payments/stripe.ts (hunk 1 above)
Comment (bbangert) [reply packages/fxa-auth-server/lib/payments/stripe.ts:3175]: In the happy path (the user is already in Stripe, with currency), there will be no Stripe calls on `expandResource` here.

File: packages/fxa-auth-server/test/local/payments/stripe.js (hunk 2 above)
Comment (bbangert) [reply packages/fxa-auth-server/test/local/payments/stripe.js:3367]: Sure. 👍

File: packages/fxa-auth-server/lib/payments/stripe.ts (hunk 3 above)
Comment (bbangert) [reply packages/fxa-auth-server/lib/payments/stripe.ts:3163]: Yep, its to catch the scenario earlier when the subscription hits its renewal period.

**Pull request merged**